File: db/test-instances.go
/*
Connect to all DNS-LG instances currently declared and check they reply
*/

package main

import (
	"github.com/mreiferson/go-httpclient" // To have HTTP timeouts
	"launchpad.net/goyaml"

	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"time"
)

type endpoints struct {
	Endpoint string
	Contact  string
	Status   string
}
type endpointsArray []endpoints

type queryObject struct {
	Versions string
	// We ignore the other fields 
}
type dnslgResponse struct {
	Query queryObject
	// We ignore the other fields 
}

type instanceTest struct {
	ok      bool
	url     string
	message string
}

var (
	verbose *bool
)

func reporter(comm chan instanceTest, over chan string, number int) {
	for i := 0; i < number; i++ {
		result := <-comm
		if *verbose || !result.ok {
			fmt.Printf("%s: %s\n", result.url, result.message)
		}
	}
	over <- "DONE"
}

func checkOne(comm chan instanceTest, url string) {
	var (
		object dnslgResponse
	)
	client := httpclient.New()
	client.ConnectTimeout = 5 * time.Second
	client.ReadWriteTimeout = 5 * time.Second
	// TODO: the format=json is redundant with the Accept header but, as of 2013-02-14, some instances still run the old code, without content negotiation
	request, err := http.NewRequest("GET", fmt.Sprintf("%s/example.org/NS?format=json", url), nil)
	if err != nil {
		comm <- instanceTest{false, url, fmt.Sprintf("Cannot create request: %s", err)}
		return
	}
	request.Header.Add("Accept", "application/json")
	request.Header.Add("User-Agent", "DNS Looking Glass Checker")
	response, err := client.Do(request)
	if err != nil {
		comm <- instanceTest{false, url, fmt.Sprintf("Cannot get: %s", err)}
		return
	}
	if response.StatusCode != http.StatusOK {
		comm <- instanceTest{false, url, fmt.Sprintf("Bad status: %s", response.Status)}
		return
	}
	defer response.Body.Close()
	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		comm <- instanceTest{false, url, fmt.Sprintf("Cannot read body: %s", err)}
		return
	}
	if len(body) == 0 {
		comm <- instanceTest{false, url, fmt.Sprintf("Got an empty body: %s", err)}
		return
	}
	err = json.Unmarshal(body, &object)
	if err != nil {
		comm <- instanceTest{false, url, fmt.Sprintf("Cannot parse the JSON result: %s", err)}
		return
	}
	comm <- instanceTest{true, url, fmt.Sprintf("OK (%d bytes) %s", len(body), object.Query.Versions)}
}

func main() {
	var (
		filename string
	)
	verbose = flag.Bool("v", false, "verbose mode, displays a line for every instance")
	flag.Parse()
	if flag.NArg() == 1 {
		filename = flag.Arg(0)
	} else {
		panic("Usage: test-instances filename.yaml\n")
	}
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	data := make([]byte, 1000000)
	count, err := file.Read(data)
	if err != nil {
		panic(err)
	}
	list := endpointsArray{}
	err = goyaml.Unmarshal(data[:count], &list)
	if err != nil {
		panic(err)
	}
	toReporter := make(chan instanceTest)
	fromReporter := make(chan string)
	go reporter(toReporter, fromReporter, len(list))

	for i := 0; i < len(list); i++ {
		go checkOne(toReporter, list[i].Endpoint)
	}
	<-fromReporter
}

File: db/test-instances.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# test-instances.py, Copyright (c) 2013, Jan-Piet Mens
# Test all instances of DNS-LG servers contained in YAML database.

import sys
import optparse
import yaml
import codecs
try:
    import json
except ImportError:
    import simplejson as json
import time
import urllib2
import socket

DBNAME = 'dns-lg.yaml'
timeout = 10

def loadf(filename):
    try:
        f = codecs.open(filename, 'r', 'utf-8')
        try:
            doc = yaml.load(f.read())
        except:
            print "Can't parse YAML in %s" % (filename)
            sys.exit(1)

        f.close()

        return doc

    except IOError, e:
        sys.stderr.write("Can't open file %s: %s\n" % (filename, e))
        sys.exit(1)

if __name__ == '__main__':

    p = optparse.OptionParser(
                usage='usage: %prog [options]',
                description='Test DNS-LG instances',
            )

    p.add_option("-f", "--database",
            action="store",
            dest="dbname",
            default=DBNAME,
            help="Path to YAML database")

    (options, args) = p.parse_args()

    sys.stdout = codecs.getwriter('utf8')(sys.stdout)

    db = loadf(options.dbname)

    for lg in db:
        if 'endpoint' in lg:
            endpoint = lg['endpoint']

            uri = "%s/example.org/NS?format=json" % (endpoint)

            try:
                socket.setdefaulttimeout(timeout)
                req = urllib2.Request(uri)
                req.add_header("Accept", "application/json")
                req.add_header("User-Agent", "DNS Looking Glass Checker")

                resp = urllib2.urlopen(req)
                content = resp.read()
            except urllib2.HTTPError, e:
                print "Can't connect to %s: %s" % (endpoint, str(e))
                continue
            except urllib2.URLError, e:
                print "Can't connect to %s: %s" % (endpoint, str(e))
                continue
            except socket.timeout:
                print "Can't connect to %s: timeout" % (endpoint)
                continue
            print endpoint
            try:
                reply = json.loads(content)
            except ValueError:
                print "\tCannot decode JSON: %s" % content
                continue
            except:
                raise

            if 'AnswerSection' in reply:
                for ans in reply['AnswerSection']:
                    print "\t%s  %s  %s" % (ans['Name'], ans['Type'], ans['Target'])

